File: c++/capacity-to-ship-packages-within-d-days.cc
class Solution {
private:
    bool canFit(vector<int>& weights, int days, int capacity) {
        int dayCount = 0;
        int curSum = 0;
        for (const auto& weight : weights) {
            if (curSum + weight > capacity) {
                curSum = 0;
                dayCount++;
            }
            curSum += weight;
        }
        dayCount++;
        return dayCount <= days;
    }
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int mn = weights[0];
        int mx = 0;
        int maxElement = weights[0];
        int minCapacity;

        for (const auto& weight : weights) {
            maxElement = max(maxElement, weight);
            mx += weight;
        }
        mn = maxElement;

        while (mn <= mx) {
            int mid = mn + (mx - mn) / 2;
            if (canFit(weights, days, mid)) {
                minCapacity = mid;
                mx = mid - 1;
            } else {
                mn  = mid + 1;
            }
        }
        return minCapacity;
    }
};

File: python/capacity-to-ship-packages-within-d-days.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        total = sum(weights)
        left = total // days
        right = total
        lowest = right

        def canShip(capacity: int):
            daysLeft = days
            index = 0
            while daysLeft and index < len(weights):
                cur = 0
                while cur <= capacity and index < len(weights):
                    if cur + weights[index] <= capacity:
                        cur += weights[index]
                        index += 1
                    else:
                        break
                daysLeft -= 1
            return index == len(weights)

        while left <= right:
            mid = (left + right) // 2
            possible = canShip(mid)
            if possible:
                lowest = min(mid, lowest)
                right = mid - 1
            else:
                left = mid + 1
        return lowest
